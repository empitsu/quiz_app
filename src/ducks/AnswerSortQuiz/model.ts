type Option = {
  optionId: number;
  originalIndex: number;
  text: string;
  selected: boolean;
};

export type SortableOptionsSet = {
  selectedOptions: Option[];
  restOptions: Option[];
};

export const sortableOptionsSet: SortableOptionsSet = {
  selectedOptions: [],
  restOptions: [],
};
