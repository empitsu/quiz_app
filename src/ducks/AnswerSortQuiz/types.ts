export const PUSH_T0_SELECTED_OPTIONS = "PUSH_T0_SELECTED_OPTIONS";
export const POP_FROM_SELECTED_OPTIONS = "POP_FROM_SELECTED_OPTIONS";
export const RESET = "RESET";
export const SHUFFLE = "SHUFFLE";
