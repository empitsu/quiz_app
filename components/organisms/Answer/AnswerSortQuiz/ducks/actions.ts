import { SortableOptionsSet } from "./model";
import {
  POP_FROM_SELECTED_OPTIONS,
  PUSH_T0_SELECTED_OPTIONS,
  RESET,
} from "./types";

export const popFromSelectedOptions = (
  removingOption: NonNullable<SortableOptionsSet["selectedOptions"][number]>
) =>
  ({
    type: POP_FROM_SELECTED_OPTIONS,
    payload: removingOption,
  } as const);

export const pushToSelectedOptions = (
  addingOption: NonNullable<SortableOptionsSet["selectedOptions"][number]>
) =>
  ({
    type: PUSH_T0_SELECTED_OPTIONS,
    payload: addingOption,
  } as const);

export const reset = (originalOptions: SortableOptionsSet["restOptions"]) =>
  ({
    type: RESET,
    payload: originalOptions,
  } as const);

export type Actions =
  | ReturnType<typeof popFromSelectedOptions>
  | ReturnType<typeof pushToSelectedOptions>
  | ReturnType<typeof reset>;
