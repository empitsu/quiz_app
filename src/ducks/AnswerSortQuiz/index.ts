export { sortableOptionsSet } from "./model";
export {
  popFromSelectedOptions,
  pushToSelectedOptions,
  reset,
} from "./actions";
export { reducer } from "./reducer";
export { isAnswerCorrect } from "./selectors";
