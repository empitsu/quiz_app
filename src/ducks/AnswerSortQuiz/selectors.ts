import { SortableOptionsSet } from "./model";

export function isAnswerCorrect(state: SortableOptionsSet): boolean {
  if (state.selectedOptions.length === 0) return false;
  return state.selectedOptions.every((option, index) => {
    return option.optionId === index + 1;
  });
}
