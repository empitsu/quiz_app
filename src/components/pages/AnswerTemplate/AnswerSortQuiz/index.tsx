import { Dispatch, useCallback, useReducer, useState } from "react";
import { reducer } from "../../../../ducks/AnswerSortQuiz";
import {
  Actions,
  popFromSelectedOptions,
  pushToSelectedOptions,
} from "../../../../ducks/AnswerSortQuiz/actions";
import { SortableOptionsSet } from "../../../../ducks/AnswerSortQuiz/model";
import { Button } from "../../../uikit/Button";
import styled from "styled-components";
import { shuffleArray } from "../../../../utils/shuffleArray";
import { AnswerResult } from "../AnswerResult";
import { Heading } from "../../../uikit/Heading";

type AnswerQuizProps = {
  title: string;
  options: {
    optionId: number;
    text: string;
  }[];
};

type EachOptionButtonProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

const StyledEmptyOptionButton = styled(Button)`
  color: transparent;
`;

function EachOptionButton({ option, dispatch }: EachOptionButtonProps) {
  const onClickAnswer = useCallback(() => {
    dispatch(pushToSelectedOptions(option));
  }, [dispatch, option]);

  if (option.selected) {
    return (
      <StyledEmptyOptionButton disabled>{option.text}</StyledEmptyOptionButton>
    );
  }

  return (
    <Button color="secondary" onClick={onClickAnswer}>
      {option.text}
    </Button>
  );
}

type SelectedOptionProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

function SelectedOption({ option, dispatch }: SelectedOptionProps) {
  const onClickSelectedOption = useCallback(() => {
    dispatch(popFromSelectedOptions(option));
  }, [dispatch, option]);

  return (
    <Button color="secondary" onClick={onClickSelectedOption}>
      {option.text}
    </Button>
  );
}

const StyledQuizTitleP = styled.p`
  text-align: center;
  margin-bottom: 30px;
`;

const StyledSelectedOptionUl = styled.ul`
  min-height: 106px;
  display: flex;
  flex-wrap: wrap;
  padding: 30px 0;
  border-top: 1px solid ${({ theme }) => theme.palettes.grey["600"]};
  border-bottom: 1px solid ${({ theme }) => theme.palettes.grey["600"]};
  margin: 30px 0;
`;

const StyledRestOptionUl = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding: 30px 0;
  margin-bottom: 50px;
`;
const StyledList = styled.li`
  list-style: none;
  margin-right: 5px;
`;

const StyledSelectionQuizWrapDiv = styled.div`
  position: relative;
`;

export function AnswerSortQuiz({ title, options }: AnswerQuizProps) {
  const officialAnswer = [...options]
    .sort()
    .map((option) => option.text)
    .join(" ");
  const defaultRestOptions = shuffleArray(options).map((option, index) => {
    return {
      ...option,
      originalIndex: index,
      selected: false,
    };
  });
  const [state, dispatch] = useReducer(reducer, {
    selectedOptions: [],
    restOptions: [...defaultRestOptions],
  });

  const [isCorrect, setIsCorrect] = useState<boolean | null>(null);

  // 正解かどうかチェックする
  const onClickAnswerBtn = useCallback(() => {
    const isCorrect =
      state.selectedOptions.length === 0
        ? false
        : state.selectedOptions.every((option, index) => {
            return option.optionId === index;
          });

    setIsCorrect(isCorrect);
  }, [state.selectedOptions]);
  return (
    <article>
      <Heading styleLevel="h3">ボタンを順番に選択してください</Heading>
      <StyledQuizTitleP>{title}</StyledQuizTitleP>
      <StyledSelectionQuizWrapDiv>
        {isCorrect === null ? (
          <div>
            <StyledSelectedOptionUl>
              {state.selectedOptions.map((option) => {
                return (
                  <StyledList key={option.optionId}>
                    <SelectedOption
                      option={option}
                      dispatch={dispatch}
                    ></SelectedOption>
                  </StyledList>
                );
              })}
            </StyledSelectedOptionUl>
            <StyledRestOptionUl>
              {state.restOptions.map((option) => {
                return (
                  <StyledList key={option.optionId}>
                    <EachOptionButton
                      option={option}
                      dispatch={dispatch}
                    ></EachOptionButton>
                  </StyledList>
                );
              })}
            </StyledRestOptionUl>
            <Button
              disabled={isCorrect !== null}
              isFullWidth
              color="secondary"
              onClick={onClickAnswerBtn}
            >
              これで回答する
            </Button>
          </div>
        ) : (
          <AnswerResult
            isCorrect={isCorrect}
            officialAnswer={officialAnswer}
          ></AnswerResult>
        )}
      </StyledSelectionQuizWrapDiv>
    </article>
  );
}
