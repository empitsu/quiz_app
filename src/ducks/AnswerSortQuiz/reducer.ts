import {
  POP_FROM_SELECTED_OPTIONS,
  PUSH_T0_SELECTED_OPTIONS,
  RESET,
} from "./types";
import { SortableOptionsSet } from "./model";
import { Actions } from "./actions";

export const reducer = (
  state: SortableOptionsSet,
  action: Actions
): SortableOptionsSet => {
  switch (action.type) {
    case POP_FROM_SELECTED_OPTIONS: {
      const targetIndex = action.payload.originalIndex;
      const currentIndex = state.selectedOptions.findIndex(
        (option) => option.optionId === action.payload.optionId
      );

      const copiedSelectedOptions = [...state.selectedOptions];
      // 選択済みからは削除する
      copiedSelectedOptions.splice(currentIndex, 1);

      // もとの選択肢には戻す
      const copiedRestOptions = [...state.restOptions];
      copiedRestOptions[targetIndex] = {
        ...action.payload,
        selected: false,
      };

      return {
        restOptions: copiedRestOptions,
        selectedOptions: copiedSelectedOptions,
      };
    }
    case PUSH_T0_SELECTED_OPTIONS: {
      const copiedSelectedOptions = [...state.selectedOptions];
      copiedSelectedOptions.push(action.payload);

      const copiedRestOptions = [...state.restOptions];
      copiedRestOptions[action.payload.originalIndex] = {
        ...copiedRestOptions[action.payload.originalIndex],
        selected: true,
      };
      return {
        restOptions: copiedRestOptions,
        selectedOptions: copiedSelectedOptions,
      };
    }
    case RESET:
      return {
        selectedOptions: [],
        restOptions: action.payload.map((option) => ({
          ...option,
          selected: false,
        })),
      };
  }
};
