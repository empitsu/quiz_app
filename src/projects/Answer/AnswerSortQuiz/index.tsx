import { Dispatch, useCallback, useContext, useReducer, useState } from "react";
import { AnswerPropStore } from "../../../contexts/AnswerProps";
import { incrementCurrentQuiz } from "../../../ducks/Answer";
import { reducer } from "../../../ducks/AnswerSortQuiz";
import {
  Actions,
  popFromSelectedOptions,
  pushToSelectedOptions,
} from "../../../ducks/AnswerSortQuiz/actions";
import { SortableOptionsSet } from "../../../ducks/AnswerSortQuiz/model";

type AnswerQuizProps = {
  title: string;
  options: SortableOptionsSet["restOptions"];
};

type EachOptionButtonProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

function EachOptionButton({ option, dispatch }: EachOptionButtonProps) {
  const onClickAnswer = useCallback(() => {
    dispatch(pushToSelectedOptions(option));
  }, [dispatch, option]);

  if (option.text === "") {
    return <p></p>;
  }

  return <button onClick={onClickAnswer}>{option.text}</button>;
}

type SelectedOptionProps = {
  option: SortableOptionsSet["restOptions"][number];
  dispatch: Dispatch<Actions>;
};

function SelectedOption({ option, dispatch }: SelectedOptionProps) {
  const onClickSelectedOption = useCallback(() => {
    dispatch(popFromSelectedOptions(option));
  }, [dispatch, option]);

  return <button onClick={onClickSelectedOption}>{option.text}</button>;
}

type ResultProps = {
  isCorrect: boolean | null;
};

function Result({ isCorrect }: ResultProps) {
  const { dispatch } = useContext(AnswerPropStore);

  if (isCorrect === null) return null;
  if (isCorrect) {
    return (
      <>
        <p>正解！</p>
        <button
          onClick={() => {
            dispatch(incrementCurrentQuiz(true));
          }}
        >
          次へ
        </button>
      </>
    );
  }
  return (
    <>
      <p>不正解！</p>
      <button
        onClick={() => {
          dispatch(incrementCurrentQuiz(false));
        }}
      >
        次へ
      </button>
    </>
  );
}

export function AnswerSortQuiz({ title, options }: AnswerQuizProps) {
  const [state, dispatch] = useReducer(reducer, {
    selectedOptions: [],
    restOptions: [...options],
  });

  const [isCorrect, setIsCorrect] = useState<boolean | null>(null);

  const onClickAnswerBtn = useCallback(() => {
    // 正解かどうかチェックする
    const isCorrect = state.selectedOptions.every((option, index) => {
      return option.optionId === index;
    });
    setIsCorrect(isCorrect);
  }, [state.selectedOptions]);
  return (
    <div>
      <p>{title}</p>
      <ul>
        {state.selectedOptions.map((option) => {
          return (
            <li key={option.optionId}>
              <SelectedOption
                option={option}
                dispatch={dispatch}
              ></SelectedOption>
            </li>
          );
        })}
      </ul>
      <ul>
        {state.restOptions.map((option) => {
          return (
            <li key={option.optionId}>
              <EachOptionButton
                option={option}
                dispatch={dispatch}
              ></EachOptionButton>
            </li>
          );
        })}
      </ul>
      <button onClick={onClickAnswerBtn}>これで回答する</button>
      <Result isCorrect={isCorrect}></Result>
    </div>
  );
}
