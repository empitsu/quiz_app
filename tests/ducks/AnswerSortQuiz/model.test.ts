import { sortableOptionsSet } from "../../../src/ducks/AnswerSortQuiz";

describe("sortableOptionsSet", () => {
  it("should create a empty model", () => {
    expect(sortableOptionsSet).toMatchSnapshot();
  });
});
