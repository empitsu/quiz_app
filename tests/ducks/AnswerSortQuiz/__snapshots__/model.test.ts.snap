// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortableOptionsSet should create a empty model 1`] = `
Object {
  "restOptions": Array [],
  "selectedOptions": Array [],
}
`;
